### PR TITLE
Test check token

### PR DESCRIPTION
**Pull request by HarithJ**
## PR Status
**READY/**

## Story ID
#153727840

## Added Migrations
NO

## Description
* Tests to make sure the user registration, authentication, profile and checks creation work as expected

## New Packages Added
None

## Related PRs 
None

## Todos
- [x] Tested and working locally
- [ ] Tested and working on development environment
- [x] Added more Unit tests (if appropriate)
- [ ] Reviewed and approved by a TTL
- [ ] Deployment to team Heroku account

## Impacted Areas in the application
* User Registration
* User login and authentication

## Known Issues (optional)
NONE

### CODE REVIEW

File: hc/accounts/tests/test_check_token.py [hunk 1]
```diff
@@ -21,8 +21,50 @@ def test_it_redirects(self):
         self.profile.refresh_from_db()
         self.assertEqual(self.profile.token, "")
 
-    ### Login and test it redirects already logged in
+    def test_login_redirects_when_already_logged_in(self):
+        """ Ensure that when a logged in user visits login page,
+        he is redirected to his profile page
+        """
+        # login
+        r = self.client.post("/accounts/check_token/alice/secret-token/")
```
Comment (johnseremba) [inline hc/accounts/tests/test_check_token.py]: use `reverse`
you can find more information about the same from the Django documentation

Comment (johnseremba) [inline hc/accounts/tests/test_check_token.py]: avoid single letter variable names.
Use more meaningful variable names

File: hc/accounts/tests/test_check_token.py [hunk 2]
```diff
@@ -21,8 +21,50 @@ def test_it_redirects(self):
...
+
+        # logging in again should redirect to checks
+        r = self.client.get("/accounts/check_token/alice/secret-token/")
+        self.assertRedirects(r, "/checks/")
+
+    def test_login_redirects_with_a_bad_token(self):
+        """ Ensure that when a bad token is provided for logging in,
+        then it should redirect to the login page 
+        """
+        r = self.client.post("/accounts/check_token/alice/bad_token/")
```
Comment (johnseremba) [inline hc/accounts/tests/test_check_token.py]: refer to the above comments

File: hc/accounts/tests/test_check_token.py [hunk 3]
```diff
@@ -21,8 +21,50 @@ def test_it_redirects(self):
...
+        
+        self.assertRedirects(r, "/accounts/login/")
+
+    def test_login_redirects_with_a_wrong_name(self):
+        """ Ensure that when a bad username is provided for logging in,
+        then it should redirect to the login page
+        """
+        r = self.client.post("/accounts/check_token/wrongName/secret-token/")
```
Comment (johnseremba) [inline hc/accounts/tests/test_check_token.py]: refer to the above comments

File: hc/accounts/tests/test_check_token.py [hunk 4]
```diff
@@ -21,8 +21,50 @@ def test_it_redirects(self):
...
+        self.assertRedirects(r, "/accounts/login/")
+
+    def test_login_redirects_to_login_with_previous_login_link(self):
+        """ Ensure that it redirects to login page,
+        when the login token is used and the user logs out
+        and tries to login again using the same token
+        """
+        # login and check if it works
+        r = self.client.post("/accounts/check_token/alice/secret-token/")
+        self.assertRedirects(r, "/checks/")
```
Comment (johnseremba) [inline hc/accounts/tests/test_check_token.py]: refer to the above comments

File: hc/accounts/tests/test_login.py [hunk 5]
```diff
@@ -19,14 +19,22 @@ def test_it_sends_link(self):
...
         self.assertEqual(mail.outbox[0].subject, 'Log in to healthchecks.io')
-        ### Assert contents of the email body
-
-        ### Assert that check is associated with the new user
+        
+        # Assert contents of the email body
+        assert re.search(r'check_token/([\w-]+)/([\w-]+)/', mail.outbox[0].body)
+
+        # Assert that check is associated with the new user
+        q = Check.objects.filter(user=user).order_by("created")
```
Comment (johnseremba) [inline hc/accounts/tests/test_login.py]: use meaningful variable names

File: hc/accounts/tests/test_login.py [hunk 6]
```diff
@@ -19,14 +19,22 @@ def test_it_sends_link(self):
         r = self.client.post("/accounts/login/", form)
         assert r.status_code == 302
 
-        ### Assert that a user was created
+        # Assert that a user was created
+        user = User.objects.get(email = form['email'])
+        assert user
```
Comment (johnseremba) [inline hc/accounts/tests/test_login.py]: use `self.assert` ...

File: hc/accounts/tests/test_login.py [hunk 7]
```diff
@@ -19,14 +19,22 @@ def test_it_sends_link(self):
...
 
         # And email sent
         self.assertEqual(len(mail.outbox), 1)
         self.assertEqual(mail.outbox[0].subject, 'Log in to healthchecks.io')
-        ### Assert contents of the email body
-
-        ### Assert that check is associated with the new user
+        
+        # Assert contents of the email body
+        assert re.search(r'check_token/([\w-]+)/([\w-]+)/', mail.outbox[0].body)
```
Comment (johnseremba) [inline hc/accounts/tests/test_login.py]: use an exclusive assert method

File: hc/accounts/tests/test_profile.py [hunk 8]
```diff
@@ -107,4 +115,29 @@ def test_it_shows_badges(self):
...
+        """Ensure that a user can create an API key and revoke it."""
+        self.client.login(username="bob@example.org", password="password")
+
+        # test that the api key is not there
+        api_key = self.bob.profile.api_key
+        self.assertEquals(api_key, "")
+
+        # create the api key
+        form = {"create_api_key": "1"}
+        r = self.client.post("/accounts/profile/", form)
```
Comment (johnseremba) [inline hc/accounts/tests/test_profile.py]: refer to the above comments

**Review by johnseremba — CHANGES_REQUESTED** [review]

File: hc/accounts/tests/test_login.py (hunk 7 above)
Comment (HarithJ) [reply hc/accounts/tests/test_login.py]: Do you mean without regex?

File: hc/accounts/tests/test_login.py (hunk 6 above)
Comment (HarithJ) [reply hc/accounts/tests/test_login.py]: Done

File: hc/accounts/tests/test_check_token.py (hunk 1 above)
Comment (HarithJ) [reply hc/accounts/tests/test_check_token.py]: Done

File: hc/accounts/tests/test_check_token.py (hunk 2 above)
Comment (HarithJ) [reply hc/accounts/tests/test_check_token.py]: Done

File: hc/accounts/tests/test_check_token.py (hunk 3 above)
Comment (HarithJ) [reply hc/accounts/tests/test_check_token.py]: Done

File: hc/accounts/tests/test_check_token.py (hunk 4 above)
Comment (HarithJ) [reply hc/accounts/tests/test_check_token.py]: Done

File: hc/accounts/tests/test_login.py (hunk 5 above)
Comment (HarithJ) [reply hc/accounts/tests/test_login.py]: Done

File: hc/accounts/tests/test_profile.py (hunk 8 above)
Comment (HarithJ) [reply hc/accounts/tests/test_profile.py]: Done

File: hc/accounts/tests/test_check_token.py (hunk 1 above)
Comment (HarithJ) [reply hc/accounts/tests/test_check_token.py]: Done

File: hc/accounts/tests/test_login.py (hunk 7 above)
Comment (HarithJ) [reply hc/accounts/tests/test_login.py]: Done

**Review by johnseremba — APPROVED** [review]
LGTM

**Pull request merged**